### PR TITLE
First Draft

### PR DESCRIPTION
**Pull request by ftzm**
I've written this simple CRUD service to get a feel for the language and how things fit together when working with ZIO. There's not a lot to it, but I'd appreciate any feedback or criticism on how to improve the project organization or how to write better/more idiomatic Scala.

I generated a few questions in the course of putting this together:
1. I'm getting my various routes via functions so that I can parameterize them with the environment type. Is there a better way to do this kind of thing?
2. I often see the function for layer creation being called "live" but also other things; is there any convention about this?
3. The DB setup here is super basic; is it more standard to set up a connection pool with Hikari or something to that effect?
4. I had planned to set up a GraphQL interface as well, but I hit a wall trying to integrate sangria with ZIO. The main issue is that field resolvers need to be of type Future, while all of my effect actions are wrapped up in ZIO types. I see that there is a `toFuture` method that might be of use here, but I see it requires something of the type `Throwable <:< Throwable` which I can make neither heads nor tails of. So two questions:
   - Are you using Sangria with ZIO-based resolvers, or falling back to plain Future?
   - Are there any plans to looks at Caliban, given that it's based on ZIO?
5. Is a there a standard approach to logging? The way I set it up seems pretty clunky; I imagine there must be a better way.
6. How long does it take before you automatically check for missing implicit imports instead of wasting 10 minutes squinting at type errors?

### CODE REVIEW

**Comment by bjarkevad** [comment]
First of all, I'm genuinely impressed with what you've written so far!
I'll try to answer your questions:

1. Typically you'd have a `type AppEnv = ZEnv with SomeService with SomeOtherService ...` that would be available to all routes. This is less granular, but it's a bit easier to work with:
```scala
type AppEnv = ZEnv // with X with Y with Z

type App[A] = ZIO[AppEnv, Throwable, A]

object ZIOHttp4sDsl extends Http4sDsl[App]

val pingRoute = {
  import ZIOHttp4sDsl._

  HttpRoutes.of[App] { case GET -> Root / "ping" =>
    Ok("pong")
  }
}

val healthRoute = {
  import ZIOHttp4sDsl._

  HttpRoutes.of[App] { case GET -> Root / "health" =>
    Ok("ok")
  }
}
```

2. The convention is kinda fuzzy to me too. I usually name layers `live`, unless there are multiple options to pick from. An example from our code base, is our test database, where we have `val standalone: ZLayer[ZEnv, Throwable, DBZ]` and `val shared: ZLayer[ZEnv with Has[DataSource], Throwable, DBZ]`, where `standalone` creates its own `DataSource`, `shared`, as the name sorta implies, can share the underlying `DataSource` with other layers. 
3. You're correct in your assumption about Hikari, and we use it too. The API is arguably garbage, but it's extremely reliable aside from that. We build doobie transactors based on the Hikari pool.
4. To answer your questions individually:
  * `Throwable <:< Throwable` is a tautology. It simply means "Prove that `Throwable` is a subtype of `Throwable`", which is trivially true in this case. It's more interesting to look at what happens if you have a custom error type:

```scala
case class MyError() extends Throwable // MyError is a subclass of Throwable
val myThing: ZIO[Any, MyError, Unit] = ZIO.unit

// Supplied by the compiler, if MyError is a subclass of Throwable
implicit val ev: MyError <:< Throwable = implicitly
myThing.toFuture // (implicit ev: MyError <:< Throwable)

val f: MyError => Throwable = myError => new Throwable("MyError")
myThing.toFutureWith(f) // (f: MyError => Throwable)
```

The `<:<` syntax, is actually a reified version of upper type bounds, so if you wanted to avoid the implicit evidence `E <:< Throwable`, you could write: 


```scala
def toFuture[R, E <: Throwable, A](zio: ZIO[R, E, A]) = zio.toFuture
``` 

* We consider Sangria fields "the end of the world", e.g. a handover to an insecure runtime. It's in no way ideal, but Sangria doesn't support custom resolvers. So yes, we use `toFuture`
  
* I've been looking at Caliban for a while, and I definitely want to migrate. I'm trying to find a way to gradually migrate, but so far I haven't found a good solution.
  
  
5. Logging is extremely legacy ridden, and not very nice to work with. Luckily it's heavily standardized, so everyone is free to use whatever logging library they wish, as long as it adheres to [SLF4J](https://en.wikipedia.org/wiki/SLF4J).
6. Implicits are probably both the worst and best feature of Scala. It's definitely the part that's hardest to learn, but it enables so many things that would otherwise be impossible or super cumbersome. It's pretty rare that I'm spending much time looking at an error, only to realize that it's caused by a missing implicit.

I hope that gives you a least some of the answers you're looking for.

File: imbibe/src/main/resources/application.conf [hunk 1]
```diff
@@ -0,0 +1,10 @@
+api {
+  host = "localhost"
+  port = 8090
+}
+db {
+  host = "localhost"
+  port = 5432
+  user = "postgres"
+  password = "hunter2"
```
Comment (bjarkevad) [inline imbibe/src/main/resources/application.conf:9]: I only see `******`

File: imbibe/src/main/scala/imbibe/Layers.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import imbibe.persistence._
+
+
+object Layers {
+  val layer0 = zio.console.Console.any ++ zio.blocking.Blocking.any ++ Config.live
+
+  val layer1 = Logging.consoleLogger ++ CocktailPersistence.layer ++ Config.live
+
+  val layers: ZLayer[Console with Blocking, Throwable, logging.Logging with HasCocktailPersistence with config.Config] =
+    layer0 >>> layer1
```
Comment (bjarkevad) [inline imbibe/src/main/scala/imbibe/Layers.scala:15]: Composing layers using `>+>` can clean this up a bit.

File: imbibe/src/main/scala/imbibe/Logger.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package imbibe
+
+import zio._
+
+package object logging {
+  type Logging = Has[Logging.Service]
+}
+
+object Logging {
+  // This is just to get a feel for effects, not my real approach to logging.
```
Comment (bjarkevad) [inline imbibe/src/main/scala/imbibe/Logger.scala:10]: Note: in our code base, we don't consider logging an effect. Not arguing that's the correct approach, but that's how it currently is.

File: imbibe/src/main/scala/imbibe/config/package.scala [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package imbibe
+
+import zio._
+
+package object config{
```
Comment (bjarkevad) [inline imbibe/src/main/scala/imbibe/config/package.scala:5]: `zio-config` makes this even nicer

File: imbibe/src/main/scala/imbibe/persistence/Persistence.scala [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+import imbibe._
+import cats.data.NonEmptyList
+
+final class CocktailPersistence(xa: Transactor[Task]) extends CocktailPersistence.Service {
+  import CocktailPersistence.SQL
+
+  def getAll(): Task[Iterable[Cocktail]] = {
+    for {
+      rows <- SQL.getAll.to[List].transact(xa).orDie
```
Comment (bjarkevad) [inline imbibe/src/main/scala/imbibe/persistence/Persistence.scala:18]: `orDie` kills the fiber, which is probably not what you want in this case

File: imbibe/src/main/scala/imbibe/persistence/Persistence.scala [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    def getAll(): Query0[Tuple2[String, String]] =
+      sql"""
+      SELECT c.name, i.name
+      FROM cocktails as c
+      JOIN ingredient_use as iu
+      ON iu.cocktail_id = c.id
+      JOIN ingredients as i
+      ON i.id = iu.ingredient_id
+      """.query[Tuple2[String, String]]
```
Comment (bjarkevad) [inline imbibe/src/main/scala/imbibe/persistence/Persistence.scala:52]: `Tuple2[A, B]` is usually written as `(A, B)`

File: imbibe/src/main/scala/imbibe/persistence/Persistence.scala [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  def xa(cfg: DbConfig) = Transactor.fromDriverManager[Task](
+    "org.postgresql.Driver", s"jdbc:postgresql://${cfg.host}:${cfg.port}/", cfg.user, cfg.password
+  )
+
+  def layer: ZLayer[Blocking with config.Config, Throwable, HasCocktailPersistence] = {
+    ZLayer.fromEffect(
+      for {
+      cfg <- config.getConfig
+    } yield new CocktailPersistence(xa(cfg.db))
+    )
```
Comment (bjarkevad) [inline imbibe/src/main/scala/imbibe/persistence/Persistence.scala:93]: could be simplified to something like `config.getConfig.map(cfg => CocktailPersistence(xa(cfg.db))).toLayer`

**Comment by ftzm** [comment]
Thanks for the detailed feedback, it's just what I was looking for! I'll see if I can't make some of the suggested improvements over the next few days.

**Pull request closed without merging**